File: ai/test/python/os_basic.py
'''
Created on Jun 18, 2016

@author: consultit
'''

import panda3d.core
from p3opensteer import OSSteerManager
from panda3d.core import LPoint3f
#
from common import app, startFramework, dataDir, getCollisionEntryFromCamera, \
            loadTerrain, mask, loadPlane
            
if __name__ == '__main__':

    app = startFramework("'one turning'")
       
    # # here is room for your own code
    
    print("create a steer manager; set root and mask to manage 'kinematic' vehicles")
    steerMgr = OSSteerManager(app.render, mask)

    print("reparent the reference node to render")
    steerMgr.get_reference_node_path().reparent_to(app.render)

    print("get a sceneNP and reparent to the reference node")
    sceneNP = loadPlane("SceneNP")
    sceneNP.reparent_to(steerMgr.get_reference_node_path())
    
    print("set sceneNP's collide mask")
    sceneNP.set_collide_mask(mask)
    
    print("create the default plug-in (attached to the reference node): 'one turning'")
    plugInNP = steerMgr.create_steer_plug_in()
    plugIn = plugInNP.node()
    
    print("get the model")
    modelNP = app.loader.load_model("eve.egg")
    modelNP.set_scale(0.25)

    print("create the steer vehicle (it is attached to the reference node) and set its position")
    vehicleNP = steerMgr.create_steer_vehicle("vehicle")
    vehicle = vehicleNP.node()
    vehicleNP.set_pos(5.0, -8.0, 0.1)
    
    print("attach the model to steer vehicle")
    modelNP.reparent_to(vehicleNP)
    
    print("add the steer vehicle to the plug-in")
    plugIn.add_steer_vehicle(vehicleNP)

    print("start the default update task for all plug-ins")
    steerMgr.start_default_update()

    print("DEBUG DRAWING: make the debug reference node paths sibling of the reference node")
    steerMgr.get_reference_node_path_debug().reparent_to(app.render)
    steerMgr.get_reference_node_path_debug_2d().reparent_to(app.aspect2d)
    print("enable debug drawing")
    plugIn.enable_debug_drawing(app.camera)

    print("toggle debug draw")
    plugIn.toggle_debug_drawing(True)
    
    # place camera
    trackball = app.trackball.node()
    trackball.set_pos(0.0, 30.0, 0.0)
    trackball.set_hpr(0.0, 20.0, 0.0)
   
    # app.run(), equals to do the main loop in C++
    app.run()

File: test/ai/c++/os_basic.cpp
/**
 * \file basic.cpp
 *
 * \date 2016-05-16
 * \author consultit
 */

#include "common.h"

int main(int argc, char *argv[])
{
	startFramework(argc, argv, "'one turning'");

	/// here is room for your own code

	cout << "create a steer manager; set root and mask to manage 'kinematic' vehicles" << endl;
	WPT(GameAIManager)steerMgr = new GameAIManager(window->get_render(), mask);

	cout << "reparent the reference node to render" << endl;
	steerMgr->get_reference_node_path().reparent_to(window->get_render());

	cout << "get a sceneNP and reparent to the reference node" << endl;
	NodePath sceneNP = loadPlane("SceneNP");
	sceneNP.reparent_to(steerMgr->get_reference_node_path());

	cout << "set sceneNP's collide mask" << endl;
	sceneNP.set_collide_mask(mask);

	cout << "create the default plug-in (attached to the reference node): 'one turning'" << endl;
	NodePath plugInNP = steerMgr->create_steer_plug_in();
	PT(OSSteerPlugIn)plugIn = DCAST(OSSteerPlugIn, plugInNP.node());

	cout << "get the model" << endl;
	NodePath modelNP = window->load_model(framework.get_models(), "eve.egg");
	modelNP.set_scale(0.25);

	cout << "create the steer vehicle (it is attached to the reference node) and set its position" << endl;
	NodePath vehicleNP = steerMgr->create_steer_vehicle("vehicle");
	PT(OSSteerVehicle)vehicle = DCAST(OSSteerVehicle, vehicleNP.node());
	vehicleNP.set_pos(5.0, -8.0, 0.1);

	cout << "attach the model to steer vehicle" << endl;
	modelNP.reparent_to(vehicleNP);

	cout << "add the steer vehicle to the plug-in" << endl;
	plugIn->add_steer_vehicle(vehicleNP);

	cout << "start the default update task for all plug-ins" << endl;
	steerMgr->start_default_update();

	cout << "DEBUG DRAWING: make the debug reference node paths sibling of the reference node" << endl;
	steerMgr->get_reference_node_path_debug().reparent_to(
			window->get_render());
	steerMgr->get_reference_node_path_debug_2d().reparent_to(
			window->get_aspect_2d());
	cout << "enable debug drawing" << endl;
	plugIn->enable_debug_drawing(window->get_camera_group());

	cout << "toggle debug draw" << endl;
	plugIn->toggle_debug_drawing(true);

	// place camera trackball (local coordinate)
	PT(Trackball)trackball = DCAST(Trackball, window->get_mouse().find("**/+Trackball").node());
	trackball->set_pos(0.0, 30.0, 0.0);
	trackball->set_hpr(0.0, 20.0, 0.0);

	// do the main loop, equals to call app.run() in python
	framework.main_loop();

	return (0);
}
